### PR TITLE
[BACKLOG-40489] Error when accessing browse files with a user with special characters

### PR DESCRIPTION
**Pull request by singletonc**
- Overload encode function to allow for disabling of double-encoding slashes, as this is not needed in all cases.

PR Set:
1. https://github.com/pentaho/pentaho-platform-plugin-common-ui/pull/1765
2. https://github.com/pentaho/pentaho-commons-gwt-modules/pull/1034
3. https://github.com/pentaho/pentaho-platform/pull/5593
4. https://github.com/pentaho/pentaho-scheduler-plugin/pull/169

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :warning: Build finished in 11m 47s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 2 test(s) failed:

<details><summary><b>.URL Encoder.URL Encoder should handle reserved URI strings</b> (click to expand)</summary><p>

```
${result.errorDetails}
```

</p></details>

<details><summary><b>.URL Encoder.URL Encoder should handle multiple occurrences of slashes</b> (click to expand)</summary><p>

```
${result.errorDetails}
```

</p></details>

  
Tests run: **5241**, Failures: **2**, Skipped: **5** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/common-ui/job/PR-1765/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: impl/client/src/main/javascript/web/util/URLEncoder.js [hunk 1]
```diff
@@ -57,7 +57,15 @@ define("common-ui/util/URLEncoder", [], function() {
 
   var O_proto = Object.prototype;
 
-  Encoder.encode = function(str, args, queryObj) {
+  /**
+   * We want to preserve the old behavior of this function here, as we have overloaded it to disable double-encoding.
+   * We do not wish to double-encode slashes in all cases, e.g. URL parameters that are paths.
+   */
+  Encoder.encode = function(str, args, queryObj){
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/util/URLEncoder.js:64]: In JavaScript, we cannot create overloads in this way. The second overload function, below, simply overwrites the first one here set on the `Encoder.encode` property. In JavaScript, functions are values.

The alternative pattern is then to have all needed parameters in a function and to explicitly check if these were specified by comparing them to `undefined`, or, in the case of a boolean property, more leniently checking if it is _nully,_ `== null` (which matches both `null` and `undefined`).

The way this is proposed, the default value of `doubleEncodeSlashes` is, in practice, inadvertently changed to `false`.

File: impl/client/src/main/javascript/web/util/URLEncoder.js [hunk 2]
```diff
@@ -70,22 +78,24 @@ define("common-ui/util/URLEncoder", [], function() {
...
+        var encodedStr = encodeURIComponent(String(item));
 
-      // Double-encode / and \ to work around Tomcat issue.
-      if(pos < pathBounds) {
-        encodedStr = encodedStr.replace(/%5C/g, "%255C").replace(/%2F/g, "%252F");
-      }
+        // Double-encode / and \ to work around Tomcat issue.
+        if (pos < pathBounds) {
+          encodedStr = encodedStr.replace(/%5C/g, "%255C").replace(/%2F/g, "%252F");
+        }
```
Comment (dcleao) [inline impl/client/src/main/javascript/web/util/URLEncoder.js:94]: The double encoding `doubleEncodeSlashes` parameter should only affect the execution of this block, and not of the entire loop, which includes the necessary `encodeURIComponent` call.

**Review by soagarwal1 — APPROVED** [review]

File: impl/client/src/main/javascript/web/util/URLEncoder.js [hunk 3]
```diff
@@ -70,22 +78,24 @@ define("common-ui/util/URLEncoder", [], function() {
...
 
-      return encodedStr;
-    });
+        return encodedStr;
+      });
+    }
```
Comment (NJtwentyone) [inline impl/client/src/main/javascript/web/util/URLEncoder.js:98]: update `URLEncoder.spec.js`

you previous made changes to https://github.com/pentaho/pentaho-platform-plugin-common-ui/blob/master/impl/client/src/test/javascript/util/URLEncoder.spec.js

If you had to make changes to the source code, then shouldn't you have to make changes to the automated test code as well?

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
update automated test class see comment - https://github.com/pentaho/pentaho-platform-plugin-common-ui/pull/1765/files#r1596759242

**Pull request closed without merging**